internal/olm/operator/catalog_source.go

```diff
@@ -0,0 +1,97 @@
+// Copyright 2020 The Operator-SDK Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package olm
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"log"
+	"strings"
+
+	operatorsv1alpha1 "github.com/operator-framework/api/pkg/operators/v1alpha1"
+	olmclient "github.com/operator-framework/operator-sdk/internal/olm/client"
+	internalregistry "github.com/operator-framework/operator-sdk/internal/olm/operator/internal"
+
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+// CatalogResources holds resources for handling CatalogSource creation
+type CatalogResources struct {
+	namespace       string
+	registryPodName string
+	registryPodAddr string
+
+	catalogsourceObj []runtime.Object
+}
+
+// initialize will initialize catalog resources with values from
+// a registry pod and validate the CatalogResources struct
+func initialize(rp *internalregistry.RegistryPod) (*CatalogResources, error) {
+	if rp == nil {
+		return nil, fmt.Errorf("registry pod is not initialized")
+	}
+
+	podName, err := rp.GetRegistryPodName()
+	if err != nil {
+		return nil, fmt.Errorf("error in getting registry pod name: %v", err)
+	}
+
+	podAddr := internalregistry.GetRegistryServiceAddr(podName, rp.Namespace)
+
+	cr := &CatalogResources{
+		namespace:       rp.Namespace,
+		registryPodName: podName,
+		registryPodAddr: podAddr,
+	}
+
+	// validate the CatalogResources struct and ensure required fields are set
+	if err := cr.validate(); err != nil {
+		return nil, fmt.Errorf("error in validating CatalogResources struct: %v", err)
+	}
+
+	return cr, nil
+}
+
+// validate will ensure that CatalogResources required fields are set
+func (cr *CatalogResources) validate() error {
+	if len(strings.TrimSpace(cr.namespace)) < 1 {
+		return errors.New("catalog namespace cannot be empty")
+	}
+	if len(strings.TrimSpace(cr.registryPodName)) < 1 {
+		return errors.New("registry pod name cannot be empty")
+	}
+
+	if len(strings.TrimSpace(cr.registryPodAddr)) < 1 {
+		return errors.New("catalog source GRPC address cannot be empty")
+	}
+
+	return nil
+}
+
+// CreateCatalogSource creates a new CatalogSource using an OLM client
+// and returns a v1alpha1 CatalogSource type and error
+func (cr *CatalogResources) CreateCatalogSource(ctx context.Context, client *olmclient.Client) (*operatorsv1alpha1.CatalogSource, error) {
```

Comment:
It is exported. where it should be used?

Reply:
This will be used in a method called `run` which will bring the registry pod up, create these additional resources such as CatalogSource, Subscription, etc.